packages/xstate-wallet/src/workflows/close-ledger-and-withdraw.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  StateNodeConfig,
+  ServiceConfig,
+  StateMachine,
+  Machine,
+  ActionFunction,
+  ActionFunctionMap,
+  DoneInvokeEvent,
+  assign,
+  ConditionPredicate
+} from 'xstate';
+import {getDataAndInvoke} from '../utils';
+import {SupportState} from '.';
+import {Store} from '../store';
+import {outcomesEqual} from '../store/state-utils';
+import {Participant, Objective, CloseLedger} from '../store/types';
+import {MessagingServiceInterface, sendDisplayMessage} from '../messaging';
+
+type WorkflowEvent = UserApproves | UserRejects | DoneInvokeEvent<CloseLedger>;
+interface UserApproves {
+  type: 'USER_APPROVES_CLOSE';
+}
+interface UserRejects {
+  type: 'USER_REJECTS_CLOSE';
+}
+export type WorkflowContext = {
+  requestId: number;
+  opponent: Participant;
+  player: Participant;
+  ledgerId?: string;
+};
+type WorkflowGuards = {
+  doesChannelIdExist: ConditionPredicate<WorkflowContext, WorkflowEvent>;
+};
+interface WorkflowActions extends ActionFunctionMap<WorkflowContext, WorkflowEvent> {
+  sendResponse: ActionFunction<WorkflowContext, WorkflowEvent>;
+  assignLedgerId: ActionFunction<WorkflowContext, DoneInvokeEvent<CloseLedger>>;
+  hideUi: ActionFunction<WorkflowContext, any>;
+  displayUi: ActionFunction<WorkflowContext, any>;
+}
+interface WorkflowServices extends Record<string, ServiceConfig<WorkflowContext>> {
+  getFinalState: (context: WorkflowContext, event: WorkflowEvent) => Promise<SupportState.Init>;
+  supportState: StateMachine<any, any, any>;
+  submitWithdrawTransaction: (context: WorkflowContext, event: WorkflowEvent) => Promise<void>;
+  createObjective: (context: WorkflowContext, event: any) => Promise<Objective>;
+}
+
+export const config: StateNodeConfig<WorkflowContext, any, any> = {
+  initial: 'waitForUserApproval',
+  states: {
+    waitForUserApproval: {
+      entry: ['displayUi'],
+      on: {
+        USER_APPROVES_CLOSE: [
+          {cond: 'doesChannelIdExist', target: 'closeLedger'},
+          'createObjective'
+        ],
+        USER_REJECTS_CLOSE: {target: 'failure'}
+      }
+    },
+
+    createObjective: {
+      invoke: {
+        src: 'createObjective',
+        onDone: {target: 'closeLedger', actions: ['assignLedgerId']}
+      }
+    },
+    closeLedger: getDataAndInvoke({src: 'getFinalState'}, {src: 'supportState'}, 'withdraw'),
```

Comment:
I was thinking about this pattern the other day. I was wondering whether an alternative would be to `getFinalState` earlier in the process and store this data on the workflow (e.g. it seems like we'd probably want to display this on the UI). It would probably also make the error handling easier in the case where you can't `getFinalState`, for whatever reason.